### PR TITLE
Add SocialAuth.beforeRedirect event

### PR DESCRIPTION
**Pull request by ADmad**
Refs #89

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -177,52 +180,134 @@ public function getUser(\Cake\Datasource\EntityInterface $profile, \Cake\Http\Se
...
+
+Upon successful authentication a `SocialAuth.afterIdentify` event is
+dispatched with the user entity. You can setup a listener for this event to
+perform required tasks. The listener can optionally return a user entity as
+event result.
+
+#### SocialAuth.beforeRedirect
+
+After the completion of authentication process before the user is redirected
+to required URL a `SocialAuth.beforeRedirect` event is trigerred. This event
```
Comment (dereuromark) [inline README.md:214]:
```suggestion
to required URL a `SocialAuth.beforeRedirect` event is triggered. This event
```

**Comment by dereuromark** [comment]
I like how now also the error flash is part of session handling.

I guess this works for me :)

**Comment by ADmad** [comment]
Can you please try out this branch in your app and let me know if everything works as expected.

**Comment by dereuromark** [comment]
Jep, works like a charm.

But the docs contain a few typos, this is the fixed class:

```php
namespace App\Event;

use ADmad\SocialAuth\Middleware\SocialAuthMiddleware;
use App\Model\Entity\User;
use Cake\Event\EventInterface;
use Cake\Event\EventListenerInterface;
use Cake\Http\ServerRequest;
use Cake\I18n\FrozenTime;
use Cake\ORM\Locator\LocatorAwareTrait;

class SocialAuthListener implements EventListenerInterface
{
    use LocatorAwareTrait;

    /**
     * @return string[]
     */
    public function implementedEvents(): array
    {
        return [
            'SocialAuth.afterIdentify' => 'afterIdentify',
            'SocialAuth.beforeRedirect' => 'beforeRedirect',
        ];
    }

    /**
     * @param \Cake\Event\EventInterface $event
     * @param \App\Model\Entity\User $user
     *
     * @return \App\Model\Entity\User
     */
    public function afterIdentify(EventInterface $event, User $user): User
    {
        // Update last login time
        $user->last_login = new FrozenTime();

        $this->getTableLocator()->get('Users')->saveOrFail($user);

        return $user;
    }

    /**
     * @param \Cake\Event\EventInterface $event
     * @param string|array $url
     * @param string $status
     * @param \Cake\Http\ServerRequest $request
     *
     * @return void
     */
    public function beforeRedirect(EventInterface $event, $url, string $status, ServerRequest $request): void
    {
        $messages = (array)$request->getSession()->read('Flash.flash');

        // Set flash message
        switch ($status) {
            case SocialAuthMiddleware::AUTH_STATUS_SUCCESS:
                $messages[] = [
                    'message' => __('You are now logged in'),
                    'key' => 'flash',
                    'element' => 'flash/success',
                    'params' => [],
                ];
                break;

            // Auth through provider failed. Details will be logged in
            // `error.log` if `logErrors` option is set to `true`.
            case SocialAuthMiddleware::AUTH_STATUS_PROVIDER_FAILURE:

                // Table finder failed to return user record. An e.g. of this is a
                // user has been authenticated through provider but your finder has
                //  condition not return an inactivated user.
            case SocialAuthMiddleware::AUTH_STATUS_FINDER_FAILURE:
                $messages[] = [
                    'message' => __('Authentication failed'),
                    'key' => 'flash',
                    'element' => 'flash/error',
                    'params' => [],
                ];
                break;
        }

        $request->getSession()->write('Flash.flash', $messages);

        // You can return a modified $redirectUrl if needed.
    }
}
```

The concrete event also should be using User instead of EntityInterface, as the last_login field is only available on the concrete object anyway.
otherwise perfect.

**Comment by ADmad** [comment]
> The concrete event also should be using User instead of EntityInterface

Since this is just example code I chose to use the interface instead of concrete class so that users can easily copy / paste to begin with. Non English users often have their users table/entity named differently.

**Comment by dereuromark** [comment]
Fair enough
I just argued from the default IDE autocomplete and stan way.
For generic docs like this still fine as is.

Nice work on the rewrite - this package now really is quite the easy to use functionality now for github (and otherwise) login.

**Pull request merged**